tests/src/test/scala/com/nvidia/spark/rapids/RapidsBufferCatalogSuite.scala

```diff
@@ -320,6 +324,125 @@ class RapidsBufferCatalogSuite extends AnyFunSuite with MockitoSugar {
+      latch.countDown()
+      catalog.registerNewBuffer(deviceBuffer)
+    }(ec)
+
+    latch.await()
+    // Let's wait for a little more, so that registerNewBuffer is likely executed first.
+    Thread.sleep(300)
+    val acquired = catalog.acquireBuffer(handle)
+    Await.ready(future, 1.second)
+    assertResult(StorageTier.DEVICE)(acquired.storageTier)
+  }
+
+  test("concurrently setting spill priority and acquiring buffer") {
+    val catalog = new RapidsBufferCatalog
+    val bufferId = MockBufferId(5)
+    val deviceBuffer = mockBuffer(bufferId, tier = DEVICE)
+    catalog.registerNewBuffer(deviceBuffer)
+    val handle = catalog.makeNewHandle(bufferId, -1)
+
+    val ec = ExecutionContext.global
+    val latch = new CountDownLatch(1)
+    val priority: Long = 54321
+
+    val future: Future[_] = Future {
+      latch.countDown()
+      deviceBuffer.setSpillPriority(priority) // some random priority
+    }(ec)
+
+    latch.await()
+    // Let's wait for a little more, so that setSpillPriority is likely executed first.
+    Thread.sleep(300)
+    val acquired = catalog.acquireBuffer(handle)
+    Await.ready(future, 1.second)
+    assertResult(priority)(acquired.getSpillPriority)
+  }
+
+  test("currently acquiring buffer with tier and removing it") {
+    val catalog = new RapidsBufferCatalog
+    val bufferId = MockBufferId(5)
+    val deviceBuffer = mockBuffer(bufferId, tier = DEVICE)
+    catalog.registerNewBuffer(deviceBuffer)
+    val handle = catalog.makeNewHandle(bufferId, -1)
+
+    val ec = ExecutionContext.global
+    val latch = new CountDownLatch(1)
+
+    val future: Future[_] = Future {
+      latch.countDown()
+      catalog.removeBufferTier(bufferId, StorageTier.DEVICE)
+    }(ec)
+
+    latch.await()
+    // Let's wait for a little more, so that removeBufferTier is likely executed first.
+    Thread.sleep(300)
+    val caught = intercept[NoSuchElementException]{
+      catalog.acquireBuffer(handle)
+    }
+    Await.ready(future, 1.second)
+    assert(caught.getMessage == s"Cannot locate buffers associated with ID: ${bufferId}")
+  }
+
+  test("currently updating tier and getting spill state") {
+    val catalog = new RapidsBufferCatalog
+    val bufferId = MockBufferId(5)
+    val deviceBuffer = mockBuffer(bufferId, tier = DEVICE)
+    val hostBuffer = mockBuffer(bufferId, tier = HOST)
+    catalog.registerNewBuffer(deviceBuffer)
+
+    val ec = ExecutionContext.global
+    val latch = new CountDownLatch(1)
+
+    val future: Future[_] = Future {
+      latch.countDown()
+      catalog.updateTiers(BufferSpill(deviceBuffer, Some(hostBuffer)))
+    }(ec)
+
+    latch.await()
+    // Let's wait for a little more, so that updateTiers is likely executed first.
+    Thread.sleep(300)
+    val isSpilled: Boolean = catalog.isBufferSpilled(bufferId, StorageTier.DEVICE)
+    Await.ready(future, 1.second)
+    assert(isSpilled)
+  }
+
+  test("currently updating tier and getting buffer meta") {
```

Comment:
`getBufferMeta` is called from the UCX shuffle. The meta is immutable and it doesn't matter if we get the "host" one vs the "device" one if I recall correctly. I don't think this is an issue.

Reply:
I don't think it's a problem of from which buffer we get the meta. The race is that the buffers could have been removed while `getBufferMeta()` is executed. It should throw instead in that case.

Reply:
So in this case, this is shuffle. The buffers are not removed all-together. They "live" somewhere, either in device/host or disk. 

Meta is something used to describe the buffer. In order for us to send the actual buffer we need to acquire it later. The idea is that this executor cached shuffle blocks that a peer is about to request. If the block is not in the catalog, we have really bad problems (because it is this executor that is currently calling `getBufferMeta` who told Spark "look I have this buffer"). 

So maybe? I just don't want to add more locks right now if we don't have a bug we are trying to fix.